### PR TITLE
Divyansh/block proposals mapdb

### PR DESCRIPTION
**Pull request by HeadTwerk**
added BlockProposalsMapDB implementation and a basic test for it

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -319,6 +319,12 @@
             <artifactId>javax.annotation-api</artifactId>
             <version>1.3.2</version>
         </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>RELEASE</version>
+            <scope>test</scope>
+        </dependency>
```
Comment (thep2p) [inline pom.xml]: This dependency has never been used, and moreover, is redundant, please remove it.

File: src/main/java/model/lightchain/BlockProposal.java [hunk 2]
```diff
@@ -13,7 +13,7 @@ public class BlockProposal extends Entity {
   private final BlockPayload payload;
 
   /**
-   * Signature of the proposer over the hash of this block.
+   * Signature of the proposer over the hash of this block.s
```
Comment (thep2p) [inline src/main/java/model/lightchain/BlockProposal.java]: Please always ensure your PR does convey inadvertent changes.

File: src/test/java/storage/BlockProposalsTest.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+  /**
+   * Test to set the last proposed block and match it against the
+   * last proposed block in database.
+   */
+  @Test
+  void setAndCheckLastBlockProposal() {
+    for (BlockProposal blockProposal : allBlockProposals) {
+      System.out.println(blockProposal);
+    }
```
Comment (thep2p) [inline src/test/java/storage/BlockProposalsTest.java]: We are not using `System.out.Println` in the test or production code. It is fine to have it for _temporary_debugging as long as it does not appear in any production code. In rare cases, if needed, we should use a logger to log the tests. However, in this test, there is no rationale behind printing block proposals, i.e., it does not add any debugging insights. Hence, please remove it.

File: src/test/java/storage/BlockProposalsTest.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.testcontainers.shaded.org.apache.commons.io.FileUtils;
+import storage.mapdb.BlockProposalsMapDb;
+import unittest.fixtures.BlockFixture;
+
+/**
+ * This class encompasses all the tests for BlockProposals implementation.
+ */
+public class BlockProposalsTest {
```
Comment (thep2p) [inline src/test/java/storage/BlockProposalsTest.java:24]: Please add tests for the following scenarios: 
1. Setting the last proposal when a proposal already exists in the `BlockProposalsMapDb` throws an `IllegalStateException`. You need to use `Assertions.assertThrows` to check the correct exception is thrown. 

2. When `BlockProposalsMapDb` is empty and 10 threads concurrently setting the last proposal, only one goes through while the rest 9 face a thrown `IllegalStateException`. 
Each thread should write a distinct block proposal. Moreover, once all threads are done, the `BlockProposalsMapDb` should be queried to get the last proposal, and the last proposal must be one of those 10 proposals, i.e., the state of last proposal should not be distorted. 

3. When`BlockProposalsMapDb` has the last proposal set, and 10 threads concurrently setting the last proposal, all 10 threads are facing a thrown `IllegalStateException`. Each thread should write a distinct block proposal. Moreover, once all threads are done, the `BlockProposalsMapDb` should be queried to get the last proposal, and the last proposal must be the one that initially was set on the database, i.e., the state of last proposal should not be distorted. 

4. When `clearLastProposal` is called on an empty data base, it does not have any effect, i.e., does not throw any exception. 

5. When `clearLastProposal` is called, it clears the last set proposal, hence getting the last proposal returns null. 

6. When`BlockProposalsMapDb` has the last proposal set, calling `clearLastProposal` multiple times (e.g., 10 times) sequentially does not have any side effects except it only cleans the last proposal (once), and the rest 9 attempts are idempotent. 

7. When`BlockProposalsMapDb` has the last proposal set, calling `clearLastProposal` multiple times (e.g., 10 times) **concurrently (each in a separate thread)** does not have any side effects except it only cleans the last proposal (once), and the rest 9 attempts are idempotent. 


💡 Note that [here](https://github.com/yhassanzadeh13/lightchain-sdp/blob/master/src/test/java/storage/BlocksTest.java#L285-L309) is a good reference example for writing concurrent tests.

**Comment by HeadTwerk** [comment]
added new test for BlockProposals

**Pull request merged**